File: NetflixRemake/app/src/main/java/br/com/titoschmidt/netflixremake/MovieActivity.java
package br.com.titoschmidt.netflixremake;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.graphics.drawable.Drawable;
import android.graphics.drawable.LayerDrawable;
import android.os.Bundle;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.appcompat.widget.Toolbar;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

import br.com.titoschmidt.netflixremake.model.Movie;
import br.com.titoschmidt.netflixremake.model.MovieDetail;
import br.com.titoschmidt.netflixremake.util.ImageDownloaderTask;
import br.com.titoschmidt.netflixremake.util.MovieDetailTask;

public class MovieActivity extends AppCompatActivity implements MovieDetailTask.MovieDetailLoader {

    private TextView txtTitulo;
    private TextView txtDesc;
    private TextView txtCast;
    private RecyclerView recyclerView;
    private MovieAdapter movieAdapter;
    private ImageView imgCover;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_movie);

        txtTitulo = findViewById(R.id.text_view_titulo);
        txtDesc = findViewById(R.id.text_view_desc);
        txtCast = findViewById(R.id.text_view_cast);
        recyclerView = findViewById(R.id.recycler_view_similar);
        imgCover = findViewById(R.id.image_view_cover);

        Toolbar toolbar = findViewById(R.id.toolbar);
        // Seta a toolbar para torná-la compatível com todos os dispositivos
        setSupportActionBar(toolbar);
        if(getSupportActionBar() != null){
            // Exibe o ícone Home
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            // Exibe o ícone o ícone Voltar
            getSupportActionBar().setHomeAsUpIndicator(R.drawable.ic_baseline_arrow_back_24);
            // Remove o título da Toolbar
            getSupportActionBar().setTitle(null);
        }

        // obs: ContextCompact possui vários métodos estáticos para tornar dispositivos compatíveis

        /* O código abaixo está acessando uma propriedade em shadows.xml

               <item android:id="@+id/cover_drawable" android:drawable="@drawable/movie_4"/>

           e trocando a propriedade desenhável (movie_4) para outra */

        LayerDrawable drawable = (LayerDrawable) ContextCompat.getDrawable(this, R.drawable.shadows);
        if(drawable != null) {
            Drawable movieCover = ContextCompat.getDrawable(this, R.drawable.movie_4);
            drawable.setDrawableByLayerId(R.id.cover_drawable, movieCover);
            //((ImageView) findViewById(R.id.image_view_cover)).setImageDrawable(drawable);
        }

        List<Movie> movies = new ArrayList<>();

        movieAdapter = new MovieAdapter(movies);

        recyclerView.setAdapter(movieAdapter);
        recyclerView.setLayoutManager(new GridLayoutManager(this,3));

        Bundle extras = getIntent().getExtras();
        if(extras != null){
            int id = extras.getInt("id");
            MovieDetailTask movieDetailTask = new MovieDetailTask(this);
            movieDetailTask.setMovieDetailLoader(this); // o listener é a MovieActivity(this) que vai escutar os eventos da Tela Inicial (MainActivity)
            movieDetailTask.execute("https://tiagoaguiar.co/api/netflix/"+id);
        }
    }
    //Button Voltar <--
    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if(item.getItemId() == android.R.id.home)
            finish();
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onResult(MovieDetail movieDetail) {
        txtTitulo.setText(movieDetail.getMovie().getTitle());
        txtDesc.setText(movieDetail.getMovie().getDescription());
        txtCast.setText(movieDetail.getMovie().getCast());

        ImageDownloaderTask imageDownloaderTask = new ImageDownloaderTask(imgCover);
        imageDownloaderTask.setShadowEnabled(true);
        imageDownloaderTask.execute(movieDetail.getMovie().getCoverUrl());

        movieAdapter.setMovies(movieDetail.getMovieSimilar());
        movieAdapter.notifyDataSetChanged();
    }

    // Classe de ViewHolder para gerenciar o layout do filme
    private class MovieHolder extends RecyclerView.ViewHolder{
        final ImageView imageViewCover;

        // Construtor
        public MovieHolder(@NonNull View itemView) {
            // itemView é o container dinâmico principal da célula movie_item, ou seja, o constraintlayout de movie_item.xml
            // que foi definido em '.inflate(R.layout.movie_item'
            super(itemView);
            // como eu inflei o movie_item.xml no MovieAdapter, eu posso obter com findViewById os elementos dentro do movie_item.xml!
            imageViewCover = itemView.findViewById(R.id.image_view_cover);
        }
    }

    // Classe de Adapter vai gerenciar a MovieHolder
    private class MovieAdapter extends RecyclerView.Adapter<MovieHolder>{

        private final List<Movie> filmes;

        private MovieAdapter(List<Movie> filmes) {
            this.filmes = filmes;
        }

        public void setMovies(List<Movie> filmes){
            this.filmes.clear();
            this.filmes.addAll(filmes);
        }

        @NonNull
        @Override
        // no OnCreate informa-se qual layout.xml será exibido e manipulado, aqui no caso será movie_item.xml
        public MovieHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            return new MovieHolder(getLayoutInflater().inflate(R.layout.movie_item_similar, parent, false)); // false, pois a raiz é o constraintlayout
        }

        @Override
        // Pega o ID do filme durante a rolagem
        public void onBindViewHolder(@NonNull MovieHolder holder, int position) {
            Movie filme = filmes.get(position);
            new ImageDownloaderTask(holder.imageViewCover).execute(filme.getCoverUrl());
        }

        @Override
        // Retorna a quantidade de filmes
        public int getItemCount() {
            return filmes.size();
        }
    }
}

File: NetflixRemake/app/src/main/java/br/com/titoschmidt/netflixremake/kotlin/MovieActivity.kt
package br.com.titoschmidt.netflixremake.kotlin

import android.graphics.drawable.Drawable
import android.graphics.drawable.LayerDrawable
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.Toast
import androidx.appcompat.app.ActionBar
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import br.com.titoschmidt.netflixremake.R
import br.com.titoschmidt.netflixremake.model.Movie
import br.com.titoschmidt.netflixremake.model.MovieDetail
import br.com.titoschmidt.netflixremake.util.MovieDetailTask
import com.bumptech.glide.Glide
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.engine.GlideException
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.target.DrawableImageViewTarget
import com.bumptech.glide.request.target.Target
import kotlinx.android.synthetic.main.activity_movie.*
import kotlinx.android.synthetic.main.activity_movie.view.*
import kotlinx.android.synthetic.main.movie_item.view.*
import kotlinx.android.synthetic.main.movie_item_similar.view.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class MovieActivity : AppCompatActivity() {

    private lateinit var movieAdapter : MovieAdapter
    private lateinit var imgCover : ImageView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_movie)

        val extras : Bundle ?= intent.extras
        if(extras != null){
            val id : Int = extras.getInt("id")
            val movieDetailTask = MovieDetailTask(this)

            retrofit().create(NetflixAPI::class.java)
                .getMovieById(id)
                .enqueue(object : Callback<MovieDetail> {
                    override fun onResponse(
                        call: Call<MovieDetail>,
                        response: Response<MovieDetail>
                    ) {
                        if(response.isSuccessful){
                            response.body()?.let { movieDetail : MovieDetail ->
                                text_view_movie_title.text = movieDetail.title
                                text_view_desc.text = movieDetail.description
                                text_view_cast.text = movieDetail.cast

                                // Carrega a imagem usando Glide
                                Glide
                                    .with(this@MovieActivity)
                                    .load(movieDetail.coverUrl)
                                    // Escuta os eventos de download
                                    // Intercepta a imagem vinda e adiciona-a no shadows.xml para receber efeito de sombreamento
                                    .listener(object : RequestListener<Drawable> {
                                        override fun onLoadFailed(
                                            e: GlideException?,
                                            model: Any?,
                                            target: Target<Drawable>?,
                                            isFirstResource: Boolean
                                        ): Boolean {
                                            return true
                                        }

                                        override fun onResourceReady(
                                            resource: Drawable?,
                                            model: Any?,
                                            target: Target<Drawable>?,
                                            dataSource: DataSource?,
                                            isFirstResource: Boolean
                                        ): Boolean {
                                            val drawable: LayerDrawable? = ContextCompat.getDrawable(baseContext, R.drawable.shadows) as LayerDrawable?
                                            drawable?.let {
                                                drawable.setDrawableByLayerId(R.id.cover_drawable, resource)
                                                (target as DrawableImageViewTarget).view.setImageDrawable(drawable)
                                            }
                                            return true
                                        }
                                    })
                                    .into(image_view_cover)
                                movieAdapter.movies.clear()
                                movieAdapter.movies.addAll(movieDetail.moviesSimilar)
                                movieAdapter.notifyDataSetChanged()
                            }
                        }
                    }

                    override fun onFailure(call: Call<MovieDetail>, t: Throwable) {
                        Toast.makeText(this@MovieActivity, t.message, Toast.LENGTH_SHORT).show()
                    }

                })

            /*
            movieDetailTask.setMovieDetailLoader { movieDetail : MovieDetail ->
                text_view_movie_title.text = movieDetail.movie.title
                text_view_desc.text = movieDetail.movie.description
                text_view_cast.text = movieDetail.movie.cast
            */
                /* Carrega a imagem da maneira antiga
                ImageDownloaderTask(image_view_cover).apply {
                    setShadowEnabled(true)
                    execute(movieDetail.movie.coverUrl)
                }
                */
            }
            //movieDetailTask.execute("https://tiagoaguiar.co/api/netflix/$id")

            // Seta a toolbar para torná-la compatível com todos os dispositivos
            setSupportActionBar(toolbar)

            supportActionBar?. let { toolbar : ActionBar ->
                toolbar.setDisplayHomeAsUpEnabled(true)
                toolbar.setHomeAsUpIndicator(R.drawable.ic_baseline_arrow_back_24)
                toolbar.setTitle(null)
            }

            val movies = arrayListOf<Movie>()
            movieAdapter = MovieAdapter(movies)
            recycler_view_similar.adapter = movieAdapter
            recycler_view_similar.layoutManager = GridLayoutManager(this,3)
    }
    //Button Voltar <--
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when(item.itemId) {
            android.R.id.home -> finish()
        }
        return super.onOptionsItemSelected(item)
    }

    private class MovieHolder(itemView : View) : RecyclerView.ViewHolder(itemView) {

        fun bind(movie : Movie){
            with(itemView) {
                //ImageDownloaderTask(itemView.image_view_cover).execute(movie.coverUrl)
                Glide
                    .with(context)
                    .load(movie.coverUrl)
                    .placeholder(R.drawable.placeholder_bg)
                    .into(image_view_similar_covers)
            }
        }
    }

    private inner class MovieAdapter(val movies : MutableList<Movie>) : RecyclerView.Adapter<MovieHolder>() {
        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MovieHolder {
            return MovieHolder(
                layoutInflater.inflate(
                    R.layout.movie_item_similar,
                    parent,
                false
                )
            )
        }

        override fun onBindViewHolder(holder: MovieHolder, position: Int) {
            val movie : Movie = movies[position]
            holder.bind(movie)
        }

        override fun getItemCount(): Int {
            return movies.size
        }
    }
}